packages/astro/src/core/render-context.ts

```diff
@@ -223,26 +222,28 @@ export class RenderContext {
 	 * API Context may be created multiple times per request, i18n data needs to be computed only once.
 	 * So, it is computed and saved here on creation of the first APIContext and reused for later ones.
 	 */
-	#i18nData?: Pick<APIContext, 'currentLocale' | 'preferredLocale' | 'preferredLocaleList'>;
-
-	get i18nData() {
-		if (this.#i18nData) return this.#i18nData;
-		const {
-			pipeline: { i18n },
-			request,
-			routeData,
-		} = this;
-		if (!i18n)
-			return {
-				currentLocale: undefined,
-				preferredLocale: undefined,
-				preferredLocaleList: undefined,
-			};
+	#currentLocale: APIContext['currentLocale'];
+	computeCurrentLocale() {
+		const { url, pipeline: { i18n }, routeData } = this;
+		if (!i18n) return;
 		const { defaultLocale, locales, strategy } = i18n;
-		return (this.#i18nData = {
-			currentLocale: computeCurrentLocale(routeData.route, locales, strategy, defaultLocale),
-			preferredLocale: computePreferredLocale(request, locales),
-			preferredLocaleList: computePreferredLocaleList(request, locales),
-		});
+		// TODO: we are making two calls to computeCurrentLocale(). In various cases, one works and the other doesn't.
+		// Ideally, we could use `renderContext.pathname` which is intended to be the one true "file-system-matchable" path.
+		// - Arsh
+		return this.#currentLocale ??= computeCurrentLocale(url.pathname, locales, strategy, defaultLocale) ?? computeCurrentLocale(routeData.route, locales, strategy, defaultLocale);
```

Comment:
As mentioned in the previously, we can't use `url.pathname` because in SSG we get a URL like `/en.html`, and we won't be able to provide the current locale.

We need to use `routeData.pathname` if it exists, and `routeData.route` as a fallback. Also. this code be cleaned up:

```js
return this.#currentLocale = computeCurrentLocale(url.pathname ?? routeData.route, locales, strategy, defaultLocale)
```